src/snark0.ml

```diff
@@ -1066,6 +1066,253 @@ module Make_basic (Backend : Backend_intf.S) = struct
+        with
+        | exception e -> Or_error.of_exn e
+        | s, x, state -> Ok (s, x, state)
+
+      let run_checked ~run ~public_input ?handler proof_system s =
+        Or_error.map (run_checked' ~run ~public_input ?handler proof_system s)
+          ~f:(fun (s, x, state) ->
+            let s', x = As_prover.run x (Checked.Runner.get_value state) s in
+            (s', x) )
+
+      let check ~run ~public_input ?handler proof_system s =
+        Or_error.is_ok
+          (run_checked' ~run ~public_input ?handler proof_system s)
+
+      let read_proving_key proof_system =
+        match proof_system.proving_key_path with
+        | Some path -> Some (Proving_key.of_string (In_channel.read_all path))
+        | None -> None
+
+      let write_proving_key proof_system key =
+        match proof_system.proving_key_path with
+        | Some path ->
+            Out_channel.write_all path ~data:(Proving_key.to_string key)
+        | None -> ()
+
+      let read_verification_key proof_system =
+        match proof_system.verification_key_path with
+        | Some path ->
+            Some (Verification_key.of_string (In_channel.read_all path))
+        | None -> None
+
+      let write_verification_key proof_system key =
+        match proof_system.verification_key_path with
+        | Some path ->
+            Out_channel.write_all path ~data:(Verification_key.to_string key)
+        | None -> ()
+
+      let prove ~run ~public_input ?proving_key ?handler proof_system s =
+        let system = R1CS_constraint_system.create () in
+        let _, _, state =
+          run_with_input ~run ~public_input ~system ?handler proof_system s
+        in
+        let {Checked.Runner.input; aux; _} = state in
+        let proving_key =
+          List.find_map_exn
+            ~f:(fun f -> f ())
+            [ (fun () -> proving_key)
+            ; (fun () -> proof_system.proving_key)
+            ; (fun () -> read_proving_key proof_system)
+            ; (fun () ->
+                let {Keypair.pk; vk} = generate_keypair ~run proof_system in
+                write_verification_key proof_system vk ;
+                Some pk ) ]
+        in
+        write_proving_key proof_system proving_key ;
+        Proof.create proving_key ~primary:input ~auxiliary:aux
+
+      let verify ~run ~public_input ?verification_key proof_system proof =
+        let input =
+          proof_system.provide_inputs (Field.Vector.create ()) public_input
+        in
+        let verification_key =
+          List.find_map_exn
+            ~f:(fun f -> f ())
+            [ (fun () -> verification_key)
+            ; (fun () -> proof_system.verification_key)
+            ; (fun () -> read_verification_key proof_system)
+            ; (fun () ->
+                failwith
+                  "Could not verify the proof; no verification key has been \
+                   provided." ) ]
+        in
+        write_verification_key proof_system verification_key ;
```

Comment:
I don't think we should be writing to disk when verify gets called (I think that can happen just once when we conditionally generate new keys)